src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -573,74 +631,87 @@ private void Poll()
+        private void ProcessResult(IImmutableSet<TopicPartition> partitionsToFetch, List<ConsumeResult<K, V>> rawResult)
         {
-            if(_log.IsDebugEnabled)
+            if (_log.IsDebugEnabled)
                 _log.Debug("Processing poll result with {0} records", rawResult.Count);
 
-            if(rawResult.IsEmpty())
+            if (rawResult.IsEmpty())
                 return;
-            
+
             var fetchedTps = rawResult.Select(m => m.TopicPartition).ToImmutableSet();
             if (!fetchedTps.Except(partitionsToFetch).IsEmpty())
                 throw new ArgumentException(
                     $"Unexpected records polled. Expected: [{string.Join(", ", partitionsToFetch.Select(p => p.ToString()))}], " +
                     $"result: [{string.Join(", ", fetchedTps.Select(p => p.ToString()))}], " +
                     $"consumer assignment: [{string.Join(", ", _consumer.Assignment.Select(p => p.ToString()))}]");
-                    
+
             //send messages to actors
-            foreach (var (stageActorRef, request) in _requests.ToTuples())
+            foreach (var (stageActorRef, request) in _requests)
             {
                 var messages = new List<ConsumeResult<K, V>>();
                 foreach (var message in rawResult)
                 {
                     var currentTp = message.TopicPartition;
-                    
+
                     if (_seekedOffset.TryGetValue(currentTp, out var seekedTpo))
                     {
                         if (message.Offset != seekedTpo.Offset)
                             throw new Exception("Seek failed, received message offset is greater than seek offset");
                         _seekedOffset = _seekedOffset.Remove(currentTp);
                     }
-                    
+
                     // If requestor is interested in consumed topic, send him consumed result
                     if (request.Topics.Contains(currentTp))
                     {
                         messages.Add(message);
                     }
                 }
-                
-                if(!messages.IsEmpty())
+
+                if (!messages.IsEmpty())
                 {
-                    stageActorRef.Tell(new KafkaConsumerActorMetadata.Internal.Messages<K, V>(request.RequestId, messages.ToImmutableList()));
+                    stageActorRef.Tell(
+                        new KafkaConsumerActorMetadata.Internal.Messages<K, V>(request.RequestId,
+                            messages.ToImmutableList()));
                     _requests = _requests.Remove(stageActorRef);
                 }
-            }                    
+            }
+        }
+
+        private void SendFailure(Exception ex, IActorRef stageActorRef)
+        {
+            stageActorRef.Tell(new Status.Failure(ex));
+            _stageActorsMap = _stageActorsMap.Where(c => !c.Value.Equals(stageActorRef)).ToImmutableDictionary();
```

Comment:
Clears "failed" stages out of the `_stageActorMap` - those stages will also call `.Fail` on themselves in response to this `Status.Failure`